proposals/p1885.md

```diff
@@ -0,0 +1,710 @@
+        and .ElementType = .CursorType.(Deref.Result) {
+    fn NewCursor[self: Self]() -> CursorType {
+      return self.begin();
+    }
+    fn Next[self: Self](cursor: CursorType*) -> Optional(ElementType) {
+      if (*cursor == self.end()) {
+        return Optional(ElementType).MakeEmpty();
+      }
+      let value: ElementType = **cursor;
+      ++*cursor;
+      return Optional(ElementType).MakeValue(value);
+    }
+  }
+}
+// used like:
+var v: Cpp.std.vector(i32) = ...
+for (i: i32 in v as CppIterate(Cpp.std.vector(i32))) { ... }
+```
+
+These two options could allow interoperability with compatible C++ type, either
+implicitly (template approach) or explicitly (adapter approach).
+
+### Inversion of control
+
+Using an Optional has limitations, discussed in the
+["alternatives considered"](#alternatives-considered) section. A possible
+approach would be to invert the control of the for loop, by having the container
+call in the `for` block for each value, passing the value as argument.
+
+This has the following advantages:
+
+-   Removes the need for an Optional, associated overheads, copies, and
+    unwrapping.
+-   No boundary checks needed at the `for` level
+-   Compatible with R-value containers
+
+Limitations:
+
+-   The context needed by the `for` body will needs to be available or provided
+-   No optimizer known to the team can reliably produce efficient code for this
+    construct.
+
+## Rationale
+
+This proposal offers a first step to support a needed way to iterate using a
+`for` loop. More specifically:
+
+-   Requires a single interface to be implemented, uses cursors instead of their
+    more complex iterator counterpart, and defines loop values as `let` by
+    default to be less error prone
+    ([Code that is easy to read, understand, and write](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/goals.md#code-that-is-easy-to-read-understand-and-write)).
+-   Makes use of
+    [Library APIs only](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/principles/library_apis_only.md).
+
+## Alternatives considered
+
+### Atomic methods for `Iterate`
+
+An option was to define methods such as `Get`, `Advance`, `IsAtEnd` on
+`Iterate`, instead of a unique `Next` methods. While being more atomic, the
+concern is that these individual methods will all need to be called separately
+while iterating, while a single call to `Next` performs these three operations
+together, providing more opportunities for optimization.
+
+Using a single operation to return an optional is also consistent with Rust’s
+approach.
+
+The drawback of this `Next` method is the fact it returns an `Optional`, though
+ways to limit or eliminate that cost seem to exist (value reuse or view for
+Optional, inversion of control on `for` block statement).
+
+### Returning `T` instead of `Optional(T)`
+
+Using `Optional(T)` is a side effect of combining operations, discussed in the
+previous section.
+
```

Comment:
I don't think this section is adding any additional value beyond the previous section.

```suggestion
```